text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,339 @@
+        if has_interface_changed(dependency_id, node_templates, cache) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+fn is_cached_implementation_still_valid(
+    id: MonoId,
+    interface_node_templates: &Map<ProgramItemId, NodeTemplate>,
+    body_node_templates: &Map<ProgramItemId, NodeTemplate>,
+    cache: &Cache)
+ -> bool {
+    
+    let node_template = body_node_templates.get(id.definition_id());
+    
+    if node_template.description() != 
+       cache.load_body_description(id.definition_id()) {
+        return false;
+    }
+
+    let node = node_template.instantiate(id.type_arguments());
+
+    for dependency_id in node.dependencies() {
+        if has_interface_changed(dependency_id, node_templates, cache) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+// Handling of cycles in dependency graph has been omitted for clarity.
+```
+
+
+## Incremental Compilation Algorithm
+With these things in place, we can construct an algorithm for building a crate incrementally. 
+
+```
+(1) Parse, Expand, Resolve whole crate
+(2) Run type inference on all program items
+(3) Build NodeTemplate map 
+(4) Add all monomorphic program items to set of needed compilation
+    artifacts
+(5) Ensure that all needed compilation artifacts are in the cache:
+    (1) Take mono-id from list of needed compilation artifacts
+    (3) Construct dependency graph to determine whether cache entry
+        can be reused
+    (4) If not, compile item
+    (5) Add item fingerprint and compilation artifacts to cache
+    (6) If new monomorphized item instances are discovered during
+        compilation, add them to list of needed compilation artifacts
+(6) Clear unreferenced entries from the cache
+(7) Create output binary
+    (a) Link cached object files into output binary, or
+    (b) Concatenate cached IR fragments, do optimization and codegen,
+        then link
+```
+
+
+## Miscellaneous Aspects
+
+### More fine-grained dependency tracking
+It would be possible to make dependency tracking aware of the kind of reference one item makes to another. If an item `A` mentions another item `B` only via some reference type (e.g. `&T`), then item `A` only needs to be updated if `B` is removed or `B` changes its 'sized-ness'. This is comparable to how forward declarations in C are handled. In the dependency graph this would mean that there are different kinds of edges that trigger for different kinds of changes to items.
+
+### Global Switches Influencing Codegen
+There are many compiler flags that change the way the generated code looks like, e.g. optimization and debuginfo levels. A simple strategy to deal with this would be to store the set of compiler flags used for building the cache and clearing the cache completely if another set of flags is used. Another option is to keep multiple caches, each for a different set of compiler flags (e.g. keeping both on disk, a 'debug build cache' and a 'release build cache').
```

Comment:
Hash the relevant flags for the subdir name? I'd expect a lot of `-C` options affect the cache, and only storing one set wouldn't help at all for some usage patterns.